spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/messaging/MessagingSpanJoiner.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright 2013-2016 the original author or authors.
+ *  
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *  
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *  
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.sleuth.instrument.messaging;
+
+import java.util.Random;
+
+import org.springframework.cloud.sleuth.Span;
+import org.springframework.cloud.sleuth.Span.SpanBuilder;
+import org.springframework.cloud.sleuth.SpanJoiner;
+import org.springframework.messaging.Message;
+
+/**
+ * Creates a {@link SpanBuilder} from {@link Message}
+ *
+ * @author Marcin Grzejszczak
+ *
+ * @since 1.0.0
+ */
+public class MessagingSpanJoiner implements SpanJoiner {
```

Comment:
Marcin think of this like gson or similar. Gson itself can serialize any
type, but to teach it a specific type, you use a jsonadapter which has a
type target parameter for the class being serialized.